packages/jest-core/src/SearchSource.ts

```diff
@@ -113,27 +111,27 @@ export default class SearchSource {
       total: allPaths.length,
     };
 
-    const testCases = Object.assign({}, this._testPathCases);
+    const testCases = this._testPathCases.slice(0);
     if (testPathPattern) {
       const regex = testPathPatternToRegExp(testPathPattern);
-      (testCases as TestPathCasesWithPathPattern).testPathPattern = (
-        path: Config.Path,
-      ) => regex.test(path);
-      (data.stats as TestPathCaseWithPathPatternStats).testPathPattern = 0;
+      testCases.push({
+        isMatch: (path: Config.Path) => regex.test(path),
+        stat: 'testPathPattern',
+      });
+      data.stats.testPathPattern = 0;
     }
 
-    const testCasesKeys = Object.keys(testCases) as Array<keyof Stats>;
-    data.tests = allPaths.filter(test =>
-      testCasesKeys.reduce<boolean>((flag, key) => {
-        const {stats} = data;
-        if (testCases[key](test.path)) {
-          stats[key] = stats[key] === undefined ? 1 : stats[key] + 1;
-          return flag && true;
+    data.tests = allPaths.filter(test => {
+      let filterResult = true;
+      for (const {isMatch, stat} of testCases) {
```

Comment:
is iterating over an array faster than `forEach` or `reduce` now?

Reply:
The reason I changed it from `reduce` to keep the code readable. Personally, it took me a few minutes to figure out exactly what it was trying to do. I find this way to be clearer.

It _is_ likely to be faster as well, because it's just less instructions for the machine to process. It doesn't have to keep passing the previous result to the next iteration, it doesn't have to do `flag && true`, etc. But those are probably relatively inconsequential.

I was just trying to keep the code as simple as possible without changing behavior so I could optimize it. If you really care I can try to benchmark both styles to see if there is a difference in this block, but I'd prefer this version.

Reply:
I can try that in isolation later as well, no worries :)